connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -836,7 +883,9 @@ public void deleteConnectorConfig(final String connName, final Callback<Created<
                     callback.onCompletion(new NotFoundException("Connector " + connName + " not found"), null);
                 } else {
                     log.trace("Removing connector config {} {}", connName, configState.connectors());
-                    configBackingStore.removeConnectorConfig(connName);
+                    if (!writeToConfigTopicAsLeader(() -> configBackingStore.removeConnectorConfig(connName))) {
+                        throw new ConnectException("Failed to remove connector configuration from config topic since worker was fenced out");
+                    }
                     callback.onCompletion(null, new Created<>(false, null));
```

Comment:
Is it OK to not invoke the `callback` in the case where we weren't leader?

Reply:
Yes, although the internal API for this is a little convoluted:
- `addRequest` accepts an action (a `Callable<Void>`) and a callback (a `Callback<Void>`)
- When requests submitted to `addRequest` are run, the callback is always invoked after they complete; if they throw an exception, it's invoked with that exception, and if they don't, it's invoked with `null` for both parameters
- The callback we pass to `addRequest` here is the result of wrapping the callback given to the `deleteConnectorConfig` method in the `forwardErrorCallback` method, which causes it to be invoked if and only if an exception is thrown when the request is run
- As a result, if we throw any exceptions from the action that we pass to `addRequest`, they're guaranteed to be passed to the callback supplied to `deleteConnectorConfig`

Although I think it's cleaner to throw exceptions instead of invoking `Callback::onCompletion` with an exception and then doing a `return null`, for consistency's sake, it's probably better to do the former, since that's the existing pattern. I'll address this first in https://github.com/apache/kafka/pull/11778 and then add it here in the subsequent rebase.

Reply:
On second thought, I think it's probably fine to leave things as they are without adding a manual invocation of `Callback::onCompletion` and a `return null`. Yes, `writeToConfigTopicAsLeader` may throw an exception, but so could writes to the config topic before changes for this KIP were made (such as [here](https://github.com/apache/kafka/blob/8e205b503a39ba8c798e3ce14cd887b66a88551c/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L835), [here](https://github.com/apache/kafka/blob/8e205b503a39ba8c798e3ce14cd887b66a88551c/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L894), and [here](https://github.com/apache/kafka/blob/8e205b503a39ba8c798e3ce14cd887b66a88551c/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L993).

If we were throwing an exception from within the body of the herder request instead of a method that the request invokes, it'd make sense to change that to instead be a manual invocation of the callback with the exception. But just calling a method that might throw an exception is fine and follows existing precedent in the code base.